devel/summarize-buildkit-output

```diff
@@ -0,0 +1,145 @@
+#!/usr/bin/env python3
+import re
+import sys
+"""
+Summarize BuildKit output by layer execution time.
+"""
+
+
+def main(file):
+    with open(file) as f:
+        id_map = get_id_map(f)
+        print_summary(id_map)
+
+
+def get_id_map(f):
+    """Return information extracted from a BuildKit output file handle.
+
+    Returns
+    -------
+    {
+        <Layer ID>: {
+            "line": The full layer start line from the output.
+            "name": A shorter version of the line that should still uniquely identify the layer.
+            "time": The build time for the layer, or None if it was cached.
+        }
+    }
+    """
+    start_pattern = re.compile(r'^(#\d+) \[(.*?)\] (.*)$')
+    cached_pattern = re.compile(r'^(#\d+) CACHED$')
+    done_pattern = re.compile(r'^(#\d+) DONE (\d+.\d+s)$')
+
+    id_map = {}
+    for line in f:
+        start_match = start_pattern.match(line)
+        if start_match:
+            layer_id = start_match.group(1)
+            if layer_id in id_map:
+                assert line == id_map[layer_id]["line"]
+                # Skip duplicate start lines.
+                continue
+
+            id_map[layer_id] = dict()
+            platform_info = get_platform_info(start_match.group(2))
+            command = start_match.group(3)
+            id_map[layer_id]["line"] = line
+            id_map[layer_id]["name"] = f"[{platform_info}] {command}"
+
+        cached_match = cached_pattern.match(line)
+        if cached_match:
+            layer_id = cached_match.group(1)
+            id_map[layer_id]["time"] = None
+
+        done_match = done_pattern.match(line)
+        if done_match:
+            layer_id = done_match.group(1)
+            if layer_id not in id_map:
+                # This can happen for tasks such as "importing cache manifest" that do not have start lines.
+                continue
+            id_map[layer_id]["time"] = get_seconds(done_match.group(2))
+
+    # Handle incomplete layers.
+    incomplete = False
+    for layer_id, info in list(id_map.items()):
+        if "time" not in info:
+            warn(f'WARNING: Build for layer {layer_id} was started but not finished.')
+            del id_map[layer_id]
+            incomplete = True
+
+    if incomplete:
+        warn("WARNING: This log is from an incomplete build.")
+
+    return id_map
+
+
+def print_summary(id_map):
+    """Print a summary of execution times per layer.
+
+    Bar charts inspired by https://alexwlchan.net/2018/ascii-bar-charts/
```

Comment:
_non-blocking_

I'm amused that blog post dubs these "ASCII" when in fact it's using Unicode block drawing chars. :upside_down_face: